Move password authentication handling into sshd/auth (fixes #394).

mik2k2 opened the pull request:
Password authentication is now completely handled in Auth. The normal
keyboard-interactive handler checks if passwords are supported and asks
for them, removing the need to override the callbacks.

Brute force throttling is removed; I'd like to base it on IP address
banning, which requires changes to the checks.
Specifically, if it's OK, I'd like to move IP and client ban checks to a new ``CheckBans``
which is called from the keyboard-interactive and public-key callbacks.
Then, a failed password could ban the IP for a configurable amount of time.
This does remove "can't be banned" status of people with a password, though.

I'm not sure, but I think timing attacks against the password are fixed:
- The hashing of the real password happens only at startup.
- The hashing of a provided password is something an attacker can do
themselves; It doesn't leak anything about the real password.
- The hash comparison is constant-time.

shazow commented on auth.go:
```diff
@@ -62,17 +66,33 @@ func NewAuth() *Auth {
 }
 
+// SetPassword anables password authentication with the given password.
+// If an emty password is geiven, disable password authentication.
```
```suggestion
// If an empty password is geiven, disable password authentication.
```

mik2k2 replied:
fixed

shazow reviewed (commented):
This is a fairly big change, but I think the reasons you've outlined make it worth it. I support your idea of refactoring the ban checks so that they can be used to throttle auth, that's a good idea.

One request: Could we change all references to "password" to be "passphrase" instead? More consistent with ssh's language, and people make better decisions about using longer strings for passphrases than passwords. :)

mik2k2 commented:
I moved the checks, ``s/assword/assphrase/`` and fixed typos. Right now, I still see these two gotchas/potentially unwanted effects:

The ban is immediate. If you mistype a password, you have to wait until it's over for a second attempt. This could incentivize either easy-to-type passphrases and short ban times (if/when they become configurable) or switching to public keys. Unless I missed a simple way to get the authentication attempt number, fixing this would require keeping track of who has how many failed attempts in which timespan, which I don't want to do unless there's a broader use for it. Note that only incorrect passwords, and not any other errors, trigger a ban.

People on a whitelist can get banned. Alone, I don't think it changes much (ops still can't ban themselves), if anything, it's a good change. However, if whitelisting is used with a passphrase, something like this could happen: 
- you're on the whitelist
- you connect
- for some reason, your key is not available (for me: my KeepassXC is locked)
-  you get a password prompt and ^D (or anything else except aborting the connection)
-  the password is wrong and you are banned
-  you make your key available
-  you are still banned and can't connect
  
Admittedly, this is a very unusual situation and not a problem at all if you ^C when you get the prompt, but I wanted to make you aware of it. This can be fixed with a bit of extra logic in the pubkey callback.

The ban time is currently hardcoded to one minute, as a tradeoff between actually bocking and being nice to people who mistype the password (although a minute certainly is quite long if you're waiting). If someone with more experience on bots has a better idea (e.g. they mostly notice they're banned and stop after 10 secs or they just continue, so it doesn't matter), it can absolutely be changed. Whoever works on #364 should probably also consider making this ban time configurable. (I might, but I'll have to read the rate-limiting code first.)

shazow commented on sshd/auth.go:
```diff
@@ -5,17 +5,26 @@ import (
+	// Given address and public key and client agent string, returns nil if the connection is not banned.
+	CheckBans(net.Addr, ssh.PublicKey, string) error
+	// Given a public key, returns nil if the connection should be allowed.
+	CheckPubkey(ssh.PublicKey) error
```
Sorry, one more nit rename: Let's call this CheckPublicKey for consistency.

mik2k2 replied:
OK

shazow commented on sshd/auth.go:
```diff
@@ -34,11 +47,32 @@ func MakeAuth(auth Auth) *ssh.ServerConfig {
+						err = auth.CheckPassphrase(answers[0])
+						if err != nil {
+							// TODO: make rate-limiting configurable
+							auth.BanAddr(conn.RemoteAddr(), time.Minute * 1)
```
I don't think we need to ban for a full minute, the idea here is to throttle brute forcing (thousands of requests per second) so just a second or two will be enough and that should resolve the UX concerns you had too.

Ideally we'd want to ban at increasing intervals, but I don't think it's worth storing additional state to achieve this right now.

shazow replied:
Also I'm down for making the ban duration configurable, at least passed in from the Host or even from `main`. We can add a cli flag later if we want.

mik2k2 replied:
OK, I was obviously thinking in the wrong order of magnitude. I set it to two seconds. With an interval that short, I agree we don't need any options to change it.

shazow reviewed (changes requested):
Looks great, thank you for putting the effort into this. You did a good job. :)

A couple of final tweaks and I think we're good to go.

shazow commented:
#364 is going to be quite different from this. It's more around rate-limiting spamming the server _after_ you're connected, so it touches several components (bandwidth-rate limiting at the socket, message-rate limiting at the room, and possibly one other I'm forgetting about). They're a bunch of magic numbers at different layers that were chosen to allow "fast human typing speed" but trip up non-trivial copypasta or people spamming aggressively. Part of it is plumbing access to these numbers, and part of it is providing a reasonable format to tweak them (or probably just allow to turn them off for fully private servers--I think there's not much point in tweaking them in practice).

shazow commented:
@mik2k2 Thank you very much! Your work is very high quality and appreciated. :)
